test-cursors/Test/Database/LMDB/Simple/Cursor/Lockstep.hs

```diff
@@ -0,0 +1,503 @@
+import           Test.Tasty                                     hiding (after)
+import           Test.Tasty.HUnit
+import           Test.Tasty.QuickCheck                          (testProperty)
+
+import           Database.LMDB.Raw
+import           Database.LMDB.Simple
+import           Database.LMDB.Simple.Cursor
+import           Database.LMDB.Simple.Internal
+
+import           Test.Database.LMDB.Simple.Cursor.Lockstep.Mock
+
+{-------------------------------------------------------------------------------
+  Main test tree
+-------------------------------------------------------------------------------}
+
+tests :: TestTree
+tests = testGroup "Lockstep" [
+    testProperty "lockstepTest" (QC.withMaxSuccess 1000 lockstepTest)
+  , testCase "labelledExamples" labelledExamples
+  ]
+
+lockstepTest :: Actions (Lockstep T) -> Property
+lockstepTest =
+  runActionsBracket
+    pT
+    (initTestEnv @IO @Word @Word)
+    cleanupTestEnv
+    (\cm tenv ->
+        readWriteTransaction (dbEnv tenv) (runCursorAsTransaction cm (db tenv))
+    )
+
+labelledExamples :: IO ()
+labelledExamples = QC.labelledExamples $
+  tagActions pT
+
+type T = CursorState Word Word
+
+pT :: Proxy T
+pT = Proxy
+
+{-------------------------------------------------------------------------------
+  Resource initialisation/cleanup
+-------------------------------------------------------------------------------}
+
+data TestEnv k v = TestEnv {
+    dbFilePath :: FilePath
+  , dbEnv      :: Environment ReadWrite
+  , db         :: Database k v
+  }
+
+initTestEnv :: MonadIO m => m (TestEnv k v)
+initTestEnv = do
+  dbFilePath <- createDbDir
+  dbEnv <- liftIO $ openEnvironment dbFilePath testLMDBLimits
+  db <- liftIO $ readWriteTransaction dbEnv (getDatabase Nothing)
+  pure TestEnv {dbFilePath, dbEnv, db}
+
+cleanupTestEnv :: MonadIO m => TestEnv k v -> m ()
+cleanupTestEnv tenv = do
+  liftIO $ closeEnvironmentBlocking (dbEnv tenv)
+  deleteDbDir (dbFilePath tenv)
+
+createDbDir :: MonadIO m => m FilePath
+createDbDir = do
+  sysTmpDir <- liftIO getTemporaryDirectory
+  liftIO $ createTempDirectory sysTmpDir "qd_cursor"
+
+deleteDbDir :: MonadIO m => FilePath -> m ()
+deleteDbDir = liftIO . removeDirectoryRecursive
+
+testLMDBLimits :: Limits
+testLMDBLimits = defaultLimits
+
+{-------------------------------------------------------------------------------
+  Model state
+-------------------------------------------------------------------------------}
+
+data CursorState k v = CursorState {
+    csMock  :: Mock k v
+  , csStats :: Stats k v
+  }
+  deriving stock (Show, Eq)
+
+initState :: CursorState k v
+initState = CursorState {
+    csMock = emptyMock
+  , csStats = initStats
+  }
+
+{-------------------------------------------------------------------------------
+  @'StateModel'@ and @'RunModel'@ instances
+-------------------------------------------------------------------------------}
+
+type RealMonad k v = CursorM k v
```

Comment:
Do we need this extra level of indirection?

Reply:
It makes swapping out/renaming the `CursorM` type easier, and the `quickcheck-lockstep` tutorial/guide suggested to do this